### ISSUE TITLE
Support XTGETTCAP/XTSETTCAP sequences

### ISSUE THREAD

**Issue by aymanbagabas** [alacritty/alacritty#7268]
Right now, Alacritty requires installing the terminal Terminfo in each machine Alacritty is running on. This is fine until we include remote sessions in the equation. On remote sessions (e.g. SSH), Terminfo databases are not always available and are sometimes not up to date. 

See: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Device-Control-functions

**Comment by kchibisov** [comment]
>Right now, Alacritty requires installing the terminal Terminfo in each machine Alacritty is running on

it's a part of ncurses project, if you want any terminfo you need ncurses, thus it'll have alacritty as a part of it. This is true for **years**. And it's spread across a lot of hosts, except for ancient ones.

And I doubt any of them support `XTGETTCAP` sequence. This sequence is mostly used by modern applications to account for new terminals or if they don't want to touch terminfo databases for whatever reason. Some apps will refuse to use `XTGETTCAP` if the `TERM` is not `xterm` for example, thus the escape is questionable.

However, we have an issue for it in https://github.com/alacritty/vte/issues/98 . It's just your motivation is not the one backed by this escape with app developers, they use it to detect non-standard features, which are not part of the terminfo, like tmux extensions.

**Comment by aymanbagabas** [comment]
> it's a part of ncurses project, if you want any terminfo you need ncurses, thus it'll have alacritty as a part of it. This is true for **years**. And it's spread across a lot of hosts, except for ancient ones.

That's true, terminfo started as a part of ncurses and it is an extension to termcap (thus the sequence name XTGETTCAP). Most hosts come with _some_ version of ncurses, not necessarily always up to date. For example, macOS comes with ncurses pre-installed but out of date and lacks terminfo databases for tmux, alacritty, and kitty. Window doesn't come with terminfo installed.

> And I doubt any of them support `XTGETTCAP` sequence.

Many terminals support the sequence, kitty, iterm2, and foot to name a few.

> This sequence is mostly used by modern applications to account for new terminals or if they don't want to touch terminfo databases for whatever reason.

Exactly! We don't need to depend on ncurses for modern applications to work correctly. Remote sessions can also benefit from this. Running modern applications on Windows will also benefit from this.

**Comment by jquast** [comment]
15 of 39 terminals tested support XTGETTCAP, https://ucs-detect.readthedocs.io/results.html#terminal-capabilities

I really like this protocol, it is allowing me to safely go "curses free" in my own code, and it's a great way to communicate additional terminal features, like Smulx, Setulc, and Ms (#8376), or the [many extended queries](https://sw.kovidgoyal.net/kitty/kittens/query_terminal/) of kitty. I hope ncurses' setupterm() function will also negotiate it someday, so that curses-linked programs can run without error, e.g.
```
$ tmux
missing or unsuitable terminal: xterm-ghostty
$ echo $?
1
```

This error exit is not necessary: ghostty fully supports XTGETTCAP and can communicate all terminal capability strings.

**Comment by j4james** [comment]
Note what the [documentation](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h4-Device-Control-functions:DCS-plus-q-Pt-ST.F95) for `XTGETTCAP` actually says:
> The string following the "q" is a list of names ... which correspond to termcap or terminfo capability names for **special keyboard keys**. ... other capabilities are intentionally not part of this interface.

I know a lot of terminals that "support" `XTGETTCAP` have basically made up their own protocol, but you aren't going to be able use `XTGETTCAP` for feature detection on terminals that are actually following the Xterm spec  (other than maybe detecting direct-color support via the special case `RGB` key).

**Comment by jquast** [comment]
Terminal software that reports full capabilities:

```
$ python dump_xtgettcap.py
contour 0.6.3-master-603c5892: Co, Setulc, Smulx, TN, bel, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, vpa
foot 1.16.2: Co, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
ghostty 1.3.2-main+4903e2821: Co, Setulc, Smulx, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
iTerm2 3.6.9: Co, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
kitty 0.46.2: Co, Setulc, Smulx, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
Rio 0.2.37: Co, Smulx, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
WezTerm 20260117-154428-05343b38: Co, Setulc, Smulx, TN, bel, blink, bold, civis, clear, cnorm, colors, cr, csr, cub, cub1, cud, cud1, cuf, cuf1, cup, cuu, cuu1, cvvis, dch, dch1, dim, dl, dl1, ech, ed, el, el1, flash, home, hpa, ich, il, il1, ind, indn, op, rc, rev, rin, ritm, rmam, rmcup, rmkx, rmso, rmul, sc, setab, setaf, sgr0, sitm, smam, smcup, smkx, smso, smul, u6, u7, u8, u9, vpa
```

Terminals reporting only Co, TN, and colors

```
xfce4-terminal 1.1.3(VTE/8400): Co, TN, colors
XTerm 406: Co, TN, colors
AbsoluteTelnet/SSH 13.16: Co, TN, colors, setab, setaf
terminator 2.1.3(VTE/8400): Co, TN, colors
termit 3.1(VTE/8400): Co, TN, colors
LXTerminal 0.4.0(VTE/7600): Co, TN, colors
mlterm 3.9.4: Co, TN, colors
GNOME Terminal 3.60.0(VTE/8400): Co, TN, colors
```

Some terminals "patch" or correct the capabilities used by ncurses. This is nice because ncurses terminfo updates are so slow to reach wide distribution, that terminal emulator authors can update capability strings without requiring accommodating changes to remote system files. ncurses terminfo has blinking inverted for `cnorm` and `cvis` compared to kitty's own distributed terminfo and XTGETTCAP response, for example.

**Comment by jquast** [comment]
I was able to successfully create https://github.com/jquast/ttyscan tool, which builds a terminfo(5) database file using XTGETTCAP under 200ms for the terminals mentioned.

Of 40 terminals tested, only 1 failed to correctly ignore XTGETTCAP, https://github.com/lxqt/qtermwidget/pull/644 and 1 failed to encode numerics correctly, https://github.com/contour-terminal/contour/pull/1948
